### ISSUE TITLE
🐞 [Bug]: Multiple impl blocks will lead to duplicate code generation (conflicting trait impls)

### ISSUE THREAD

**Issue by VivekYadav7272** [ozgunozerk/state-shift#20]
### What happened?

Say you have some code:
``` rust
use state_shift::type_state;

#[state_shift::type_state(state_slots=3, default_state=Uninit)]
pub struct FooBuilder {
    foo: i32,
    bar: i32,
    baz: i32,
}

#[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
impl FooBuilder {
 // .. some code here
}

#[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
impl FooBuilder {
    // .. some code here
}
```

This is one of the things that I like about Rust which is that you can logically separate your methods into different blocks. However, due to how the crate generates state structs, having two blocks will lead them to be redefined.
Here's the error:
``` rust
10 | #[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
   | --------------------------------------------------------- previous definition of the module `sealed_foo_builder` here
...
13 | #[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `sealed_foo_builder` redefined here
   |
   = note: `sealed_foo_builder` must be defined only once in the type namespace of this module
   = note: this error originates in the attribute macro `state_shift::states` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0428]: the name `SealerFooBuilder` is defined multiple times
  --> src/main.rs:13:1
   |
10 | #[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
   | --------------------------------------------------------- previous definition of the trait `SealerFooBuilder` here
...
13 | #[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `SealerFooBuilder` redefined here
   |
   = note: `SealerFooBuilder` must be defined only once in the type namespace of this module
   = note: this error originates in the attribute macro `state_shift::states` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0428]: the name `FooBuilderUninit` is defined multiple times
  --> src/main.rs:13:1
   |
10 | #[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
   | --------------------------------------------------------- previous definition of the type `FooBuilderUninit` here
...
13 | #[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FooBuilderUninit` redefined here
   |
   = note: `FooBuilderUninit` must be defined only once in the type namespace of this module
   = note: this error originates in the attribute macro `state_shift::states` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0428]: the name `FooBuilderFooInit` is defined multiple times
  --> src/main.rs:13:1
   |
10 | #[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
   | --------------------------------------------------------- previous definition of the type `FooBuilderFooInit` here
...
13 | #[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FooBuilderFooInit` redefined here
   |
   = note: `FooBuilderFooInit` must be defined only once in the type namespace of this module
   = note: this error originates in the attribute macro `state_shift::states` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0428]: the name `FooBuilderBarInit` is defined multiple times
  --> src/main.rs:13:1
   |
10 | #[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
   | --------------------------------------------------------- previous definition of the type `FooBuilderBarInit` here
...
13 | #[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FooBuilderBarInit` redefined here
   |
   = note: `FooBuilderBarInit` must be defined only once in the type namespace of this module
   = note: this error originates in the attribute macro `state_shift::states` (in Nightly builds, run with -Z macro-backtrace for more info)
```
... and it goes on.

### Expected behavior

In my opinion, trying to define the possible states on impl block from DX POV doesn't make sense anyways. Impl blocks aren't necessarily located close to the struct definition which makes it slightly annoying if you have to go to the impl block to find a piece of information you'd expect to be part of struct itself. But maybe I'm saying that because I'm used to the boilerplate typestate pattern. Well anyways, if we define it on the struct, then you can create traits and state structs and impl Protector trait on them without the fear of duplicate generation since structs can't be defined twice, and then the attribute macro on the impl blocks can just care about introducing the generic parameters onto the block.

I haven't written proc macros though, so I'm not sure how you can manage to introduce generic parameters without needing the same #[states(..)] macro again, which makes it cumbersome.

**Comment by ozgunozerk** [comment]
Hey Vivek! This was discovered couple of days ago, I think I fixed it already.

 You can check out the new release `v1.1.0`

Will close it because I think it is duplicate of #16.

Feel free to re-open this if you think the issue persists in `v1.1.0` 👍

**Comment by ozgunozerk** [comment]
Oh no this is already on `v1.1.0` hahah, my bad

**Comment by ozgunozerk** [comment]
Yeah you cannot do that:
```
#[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
impl FooBuilder {
 // .. some code here
}

#[state_shift::states(Uninit, FooInit, BarInit, BazInit)]
impl FooBuilder {
    // .. some code here
}
```

I think there are no good ways to solve this. Is there specific need to separate the impl block into two?

You can merge them into a single one?

**Comment by VivekYadav7272** [comment]
Well there's no strong reasons, it's mostly done for clarity and manageability of code. More importantly it's a restriction that's implicit and one may want to know about this added "cost" beforehand, so atleast if it could be documented that'd be nice.
>the reason is, states macro generates many marker structs behind the scenes. And from what I know, proc macros cannot be aware of other proc macros (unless they are on the same item).
Yes that is true. However, like I suggested (and this is only my opinion) - we shouldn't be declaring the possible states on an impl block anyways. Semantically, it should be on the struct itself. That way this problem could be solved because then you can delegate the declaration of marker state structs in that macro only. But yes, without changing anything else, these attribute macros would still be required on each impl block but atleast multiple impl blocks wouldn't clash.

**Comment by jkneer** [comment]
Hi,

To me it would also be more plausible to have the states declaration on the struct. Just by how we refer the macros in this discussion shows that is a "declaration/definition" not an implementation.

**Comment by ozgunozerk** [comment]
> Yes that is true. However, like I suggested (and this is only my opinion) - we shouldn't be declaring the possible states on an impl block anyways. Semantically, it should be on the struct itself. That way this problem could be solved because then you can delegate the declaration of marker state structs in that macro only. But yes, without changing anything else, these attribute macros would still be required on each impl block but atleast multiple impl blocks wouldn't clash.

I agree, these are good points. I can do smt like that:
```rust
#[type_state(state_slots=3, default_state=Uninit)]
#[states(Uninit, FooInit, BarInit, BazInit)]
pub struct FooBuilder {
    foo: i32,
    bar: i32,
    baz: i32,
}

#[stateful_methods] /* we have to have a macro that wraps the impl block. I haven't spent time on the name */ 
impl FooBuilder {
   #[require(BarInit)]
   // some method

   /* redacted */
}
```

In fact, this was my initial design. Let me explain why did I abandon it:
- `require` and `switch_to` macros need to have access the name of the struct, and the generics/lifetimes provided to `impl` statement itself. 
- Unfortunately, these macros only have access to the content of the method. They cannot access the `impl` block's content. 
- that's why we need another macro that wraps the whole `impl` block, and pass the above mentioned information to `require` and `switch_to`
- that macro we need on top of `impl` block feels weird imo, and it doesn't need to take any argument.
- so I thought, I could move the `states` macro on top of the `impl` block, and merge the functionalities under one macro
- this also reduced the amount of macros needs to be used

However, your concerns and preferences are 100% valid.

So, @VivekYadav7272 and @jkneer, would love to hear your opinions on this. If you both simply prefer the code snippet I put above, it is a fairly easy change for me to implement. 

If you have other ideas, I'm all ears.

**Comment by jkneer** [comment]
I like that better. Also can states and type_state be one macro?

**Comment by ozgunozerk** [comment]
It can tbh, but the arguments would look too convoluted imo, and will occupy 2 lines anyway in most cases i guess

**Comment by VivekYadav7272** [comment]
Oh, I was under the impression that such design would be a lot more difficult because the impl block macros i.e require and switch_to, would need to know the names of the states and that can only happen with the macro that sits on the impl block. 
If you are suggesting that the code-snippet provided by you is feasible, I would 100% prefer it. It may be a little confusing as to why we require one on the impl block, but I think there's enough weird things like that about proc-macros that it doesn't make it any weirder than they are in general.
Also, not the place, but if you could help me understand a little bit better why you think this is not that hard to fix it would be nice, since again, I thought the state names being mentioned on top of the impl block was necessary.

**Comment by jkneer** [comment]
My reasoning for asking if it can be a single marco on the struct: If it is a single macro you get lints to guide you through the full setup. If it is a second macro, you get no reminder of the missing second macro and it is a cryptic compile time error.

**Comment by ozgunozerk** [comment]
@VivekYadav7272 

> i.e require and switch_to, would need to know the names of the states and that can only happen with the macro that sits on the impl block.

Actually no, `require` and `switch_to` needs to know the states that are supplied to them as arguments. They don't need the full set of the states. 

They only need to know the following:
- name of the struct
- lifetimes / generics provided to the impl block
- where clauses

That's why we need a wrapper on top of the impl block, but we don't need the full states fed to each `require` and `switch_to`. 

> Also, not the place, but if you could help me understand a little bit better why you think this is not that hard to fix it would be nice,

Yeah sure. `states` macro as of now has 2 responsibilities:
- providing the info mentioned above to the `require` and `switch_to`
- creating the marker structs and traits

Since these 2 items are quite distinct tasks, I reckon the separation will be quite easy.

**Comment by ozgunozerk** [comment]
@jkneer 

I see your points. The more I think on it, the more I felt closer to your suggestion.

The things we are discussing here will significantly change the API. This is perfectly ok as of now, since we are still on the early phases :)

However, let's think on it a couple of days more before committing all these, because they will be quite BREAKING changes.

Here is what we have for the comparison:

```rust
#[type_state(
    states = (Uninit, FooInit, BarInit, BazInit)
    state_slots = 2, 
    default_state = Uninit, 
)]
pub struct FooBuilder {
    foo: i32,
    bar: i32,
    baz: i32,
}
```

vs.

```rust 
#[type_state(state_slots=3, default_state=Uninit)]
#[states(Uninit, FooInit, BarInit, BazInit)]
pub struct FooBuilder {
    foo: i32,
    bar: i32,
    baz: i32,
}
```

I'll start implementing these tomorrow, will let you know in here

**Comment by jkneer** [comment]
I like that and seeing this I wonde if every slot should be able to have its own default?

Then the syntax could become something like this:
```rust
#[type_state(
    states = (Uninit, FooInit, BarInit, BazInit)
    defaults = (Uninit, Uninit), 
)]
pub struct FooBuilder {
    foo: i32,
    bar: i32,
    baz: i32,
}
```

**Comment by ozgunozerk** [comment]
That's terrific! 
Simpler and more expressive 💯
